File: src/review.ts
import { getAddress } from 'ethers';
import { TxCompilerError } from './errors.js';
import type { FeeParams, FeeReview, PreparedTransaction, TransactionReview } from './types.js';

const ERC20_TRANSFER_SELECTOR = 'a9059cbb';
const ERC20_TRANSFER_CALLDATA_HEX_LEN = 136;
const ERC20_ADDRESS_PADDING = '0'.repeat(24);
const HEX_RE = /^[0-9a-fA-F]+$/;
const NON_NEG_INT_RE = /^(0|[1-9]\d*)$/;

/**
 * Build a human review summary for the currently supported EVM flows.
 *
 * Native transfers use top level fields directly. ERC-20 transfers derive the
 * actual recipient and amount from calldata so confirmation screens reflect the
 * on chain effect rather than the transaction envelope.
 */
export function review(prepared: PreparedTransaction): TransactionReview {
  const { chain, txType, from, to, valueWei, nonce, chainId, fee } = prepared;

  if (chain !== 'ethereum') {
    throw new TxCompilerError('UNSUPPORTED_CHAIN', `${chain} review not supported yet`);
  }

  validateEvmAddress(from, 'from');

  const normalizedChainId = requirePositiveChainId(chainId);
  const normalizedNonce = requireNonce(nonce);

  let recipient: string;
  let amount: string;
  let tokenContract: string | null;

  if (txType === 'TRANSFER_NATIVE') {
    validateEvmAddress(to, 'to');
    tokenContract = resolveTokenContract(prepared);
    recipient = to;
    amount = requireUnsignedDecimal(valueWei, 'valueWei');
  } else if (txType === 'TRANSFER_TOKEN') {
    tokenContract = resolveTokenContract(prepared);
    if (prepared.data == null) {
      throw new TxCompilerError('INVALID_CALLDATA', 'EVM token transfer missing calldata (data)');
    }

    const decoded = decodeErc20TransferData(prepared.data);
    validateEvmAddress(decoded.recipient, 'recipient');
    recipient = decoded.recipient;
    amount = decoded.amount;
  } else {
    throw new TxCompilerError('UNSUPPORTED_TX_TYPE', `${txType} review not supported yet`);
  }

  return {
    chain,
    txType,
    from,
    recipient,
    amount,
    tokenContract,
    nonce: normalizedNonce,
    chainId: normalizedChainId,
    fee: buildFeeReview(fee),
  };
}

/**
 * Decode the recipient and amount from ERC-20 transfer calldata.
 */
function decodeErc20TransferData(data: string): { recipient: string; amount: string } {
  const hex = data.startsWith('0x') ? data.slice(2) : data;

  if (!HEX_RE.test(hex)) {
    throw new TxCompilerError(
      'INVALID_CALLDATA',
      'ERC-20 transfer calldata contains non-hex characters',
    );
  }

  if (hex.length !== ERC20_TRANSFER_CALLDATA_HEX_LEN) {
    throw new TxCompilerError(
      'INVALID_CALLDATA',
      `ERC-20 transfer calldata must be exactly 68 bytes (got ${String(hex.length / 2)})`,
      { length: hex.length, expected: ERC20_TRANSFER_CALLDATA_HEX_LEN },
    );
  }

  const selector = hex.slice(0, 8).toLowerCase();
  if (selector !== ERC20_TRANSFER_SELECTOR) {
    throw new TxCompilerError(
      'INVALID_CALLDATA',
      `Expected ERC-20 transfer selector ${ERC20_TRANSFER_SELECTOR}, got ${selector}`,
    );
  }

  const addressPadding = hex.slice(8, 32);
  if (addressPadding !== ERC20_ADDRESS_PADDING) {
    throw new TxCompilerError(
      'INVALID_CALLDATA',
      'ERC-20 transfer address has non-zero upper bytes',
    );
  }

  const recipient = '0x' + hex.slice(32, 72);
  const amountHex = hex.slice(72, 136);

  return {
    recipient,
    amount: BigInt('0x' + amountHex).toString(),
  };
}

/**
 * Resolve the token contract that will actually be executed on chain.
 */
function resolveTokenContract(prepared: PreparedTransaction): string | null {
  if (prepared.txType !== 'TRANSFER_TOKEN') {
    if (prepared.tokenContract != null) {
      throw new TxCompilerError(
        'INVALID_PAYLOAD',
        'EVM TRANSFER_NATIVE must not include tokenContract',
      );
    }

    return null;
  }

  validateEvmAddress(prepared.to, 'to');

  if (prepared.tokenContract == null) {
    return prepared.to;
  }

  validateEvmAddress(prepared.tokenContract, 'tokenContract');

  if (prepared.tokenContract.toLowerCase() !== prepared.to.toLowerCase()) {
    throw new TxCompilerError(
      'INVALID_PAYLOAD',
      'EVM token transfer tokenContract must match the transaction target',
      {
        to: prepared.to,
        tokenContract: prepared.tokenContract,
      },
    );
  }

  return prepared.to;
}

/**
 * Summarize fee fields into confirmation friendly review values.
 */
function buildFeeReview(fee: FeeParams): FeeReview {
  if (fee.mode === 'EIP1559') {
    const gasLimit = requireFeeField(fee.gasLimit, 'gasLimit');
    const maxFeePerGas = requireFeeField(fee.maxFeePerGas, 'maxFeePerGas');
    const maxPriorityFeePerGas = requireFeeField(
      fee.maxPriorityFeePerGas,
      'maxPriorityFeePerGas',
    );
    const baseFeePerGas = optionalUnsignedDecimal(fee.baseFeePerGas, 'baseFeePerGas');

    return {
      mode: fee.mode,
      estimatedMaxCost: (BigInt(gasLimit) * BigInt(maxFeePerGas)).toString(),
      gasLimit,
      maxFeePerGas,
      maxPriorityFeePerGas,
      baseFeePerGas,
    };
  }

  if (fee.mode === 'LEGACY') {
    const gasLimit = requireFeeField(fee.gasLimit, 'gasLimit');
    const gasPrice = resolveLegacyGasPrice(fee);

    return {
      mode: fee.mode,
      estimatedMaxCost: (BigInt(gasLimit) * BigInt(gasPrice)).toString(),
      gasLimit,
      gasPrice,
    };
  }

  throw new TxCompilerError(
    'UNSUPPORTED_FEE_MODE',
    `${fee.mode} fee review invalid for ethereum chain`,
  );
}

function validateEvmAddress(address: string, field: string): void {
  try {
    getAddress(address);
  } catch {
    throw new TxCompilerError('INVALID_ADDRESS', `Invalid EVM address in ${field}: ${address}`, {
      field,
      address,
    });
  }
}

function requirePositiveChainId(chainId: number | null): number {
  if (chainId == null || !Number.isSafeInteger(chainId) || chainId < 1) {
    throw new TxCompilerError('INVALID_PAYLOAD', 'EVM chainId must be a positive integer', {
      chainId,
    });
  }

  return chainId;
}

function requireNonce(nonce: string | null): string {
  if (nonce == null) {
    throw new TxCompilerError('INVALID_PAYLOAD', 'EVM transactions require nonce');
  }

  return requireUnsignedDecimal(nonce, 'nonce');
}

function requireUnsignedDecimal(value: string, field: string): string {
  if (!NON_NEG_INT_RE.test(value)) {
    throw new TxCompilerError('INVALID_AMOUNT', `${field} must be a non-negative integer`, {
      field,
      value,
    });
  }

  return value;
}

function optionalUnsignedDecimal(value: string | null | undefined, field: string): string | undefined {
  if (value == null) return undefined;
  return requireUnsignedDecimal(value, field);
}

function requireFeeField(value: string | null | undefined, field: string): string {
  if (value == null) {
    throw new TxCompilerError('MISSING_FEE_PARAMS', `${field} is required for fee review`);
  }

  return requireUnsignedDecimal(value, field);
}

function resolveLegacyGasPrice(fee: FeeParams): string {
  if (fee.gasPrice != null) {
    return requireUnsignedDecimal(fee.gasPrice, 'gasPrice');
  }

  if (fee.maxFeePerGas != null) {
    return requireUnsignedDecimal(fee.maxFeePerGas, 'maxFeePerGas');
  }

  throw new TxCompilerError('MISSING_FEE_PARAMS', 'LEGACY requires gasPrice or maxFeePerGas');
}

File: src/review.rs
//! Derive a human-reviewable summary from a prepared transaction.

use alloy_primitives::U256;

use crate::constants::{ERC20_TRANSFER_CALLDATA_HEX_LEN, ERC20_TRANSFER_SELECTOR_HEX};
use crate::error::{TxCompilerError, TxCompilerErrorCode};
use crate::types::{
    Chain, FeeMode, FeeParams, FeeReview, PreparedTransaction, TransactionReview, TxType,
};

/// Produce a human-reviewable summary of `prepared`.
///
/// For EVM token transfers the actual recipient and amount are decoded from
/// the calldata, so the review reflects what will execute on-chain rather
/// than the raw envelope fields.
pub fn review(prepared: &PreparedTransaction) -> Result<TransactionReview, TxCompilerError> {
    let (recipient, amount) = match (prepared.tx_type, prepared.chain) {
        (TxType::TransferNative, _) | (TxType::TransferToken, Chain::Tron) => {
            (prepared.to.clone(), prepared.value_wei.clone())
        }
        (TxType::TransferToken, Chain::Ethereum) => {
            let data = prepared.data.as_deref().ok_or_else(|| {
                TxCompilerError::new(
                    TxCompilerErrorCode::InvalidCalldata,
                    "EVM token transfer missing calldata (data)",
                )
            })?;
            decode_erc20_transfer_data(data)?
        }
    };

    Ok(TransactionReview {
        chain: prepared.chain,
        tx_type: prepared.tx_type,
        from: prepared.from.clone(),
        recipient,
        amount,
        token_contract: prepared.token_contract.clone(),
        nonce: prepared.nonce.clone(),
        chain_id: prepared.chain_id,
        fee: build_fee_review(&prepared.fee),
    })
}

fn decode_erc20_transfer_data(data: &str) -> Result<(String, String), TxCompilerError> {
    let hex_str = data.strip_prefix("0x").unwrap_or(data);

    if !is_hex(hex_str) {
        return Err(TxCompilerError::new(
            TxCompilerErrorCode::InvalidCalldata,
            "ERC-20 transfer calldata contains non-hex characters",
        ));
    }

    if hex_str.len() != ERC20_TRANSFER_CALLDATA_HEX_LEN {
        return Err(TxCompilerError::with_details(
            TxCompilerErrorCode::InvalidCalldata,
            format!(
                "ERC-20 transfer calldata must be exactly 68 bytes (got {})",
                hex_str.len() / 2
            ),
            serde_json::json!({
                "length": hex_str.len(),
                "expected": ERC20_TRANSFER_CALLDATA_HEX_LEN,
            }),
        ));
    }

    let selector = hex_str[0..8].to_ascii_lowercase();
    if selector != ERC20_TRANSFER_SELECTOR_HEX {
        return Err(TxCompilerError::with_details(
            TxCompilerErrorCode::InvalidCalldata,
            "Calldata does not start with the ERC-20 transfer selector",
            serde_json::json!({
                "expected": ERC20_TRANSFER_SELECTOR_HEX,
                "actual": selector,
            }),
        ));
    }

    // Enforce the 24-char left zero-pad even on this code path. `validate`
    // already does this, but `review` may be called on a
    // `PreparedTransaction` constructed by any other path (direct serde
    // deserialise from a hostile source, for instance). Without this guard
    // a crafted calldata would silently truncate the high bytes of the
    // address word and display a different recipient than what the signer
    // will actually sign for.
    let address_word = &hex_str[8..72];
    if !address_word[..24].bytes().all(|b| b == b'0') {
        return Err(TxCompilerError::new(
            TxCompilerErrorCode::InvalidCalldata,
            "ERC-20 transfer recipient word must be left-zero-padded",
        ));
    }
    let recipient_hex = &address_word[24..];
    let recipient = format!("0x{recipient_hex}");

    let amount_hex = &hex_str[72..136];
    let amount_u256 = U256::from_str_radix(amount_hex, 16).map_err(|_| {
        TxCompilerError::new(
            TxCompilerErrorCode::InvalidCalldata,
            "ERC-20 transfer amount is not valid hex",
        )
    })?;

    Ok((recipient, amount_u256.to_string()))
}

fn build_fee_review(fee: &FeeParams) -> FeeReview {
    match fee.mode {
        FeeMode::Eip1559 => {
            let gas_limit = fee.gas_limit.clone().unwrap_or_else(|| "0".into());
            let max_fee = fee.max_fee_per_gas.clone().unwrap_or_else(|| "0".into());
            let estimated = multiply_decimal(&gas_limit, &max_fee);
            FeeReview {
                mode: FeeMode::Eip1559,
                estimated_max_cost: Some(estimated),
                gas_limit: Some(gas_limit),
                gas_price: None,
                max_fee_per_gas: Some(max_fee),
                max_priority_fee_per_gas: fee.max_priority_fee_per_gas.clone(),
                base_fee_per_gas: fee.base_fee_per_gas.clone(),
                tron_fee_limit: None,
            }
        }
        FeeMode::Legacy => {
            let gas_limit = fee.gas_limit.clone().unwrap_or_else(|| "0".into());
            let gas_price = fee
                .gas_price
                .clone()
                .or_else(|| fee.max_fee_per_gas.clone())
                .unwrap_or_else(|| "0".into());
            let estimated = multiply_decimal(&gas_limit, &gas_price);
            FeeReview {
                mode: FeeMode::Legacy,
                estimated_max_cost: Some(estimated),
                gas_limit: Some(gas_limit),
                gas_price: Some(gas_price),
                max_fee_per_gas: None,
                max_priority_fee_per_gas: None,
                base_fee_per_gas: None,
                tron_fee_limit: None,
            }
        }
        FeeMode::Tron => FeeReview {
            mode: FeeMode::Tron,
            estimated_max_cost: fee.el.clone(),
            gas_limit: None,
            gas_price: None,
            max_fee_per_gas: None,
            max_priority_fee_per_gas: None,
            base_fee_per_gas: None,
            tron_fee_limit: fee.el.clone(),
        },
    }
}

fn multiply_decimal(a: &str, b: &str) -> String {
    // U256 × U256; sufficient for any realistic gas × gas-price product.
    let Ok(lhs) = U256::from_str_radix(a, 10) else {
        return "0".into();
    };
    let Ok(rhs) = U256::from_str_radix(b, 10) else {
        return "0".into();
    };
    lhs.saturating_mul(rhs).to_string()
}

fn is_hex(s: &str) -> bool {
    !s.is_empty() && s.bytes().all(|b| b.is_ascii_hexdigit())
}
